src/actions/post_build.rs

```diff
@@ -47,13 +47,20 @@ pub struct PostBuildHandler {
 // is not object-safe).
 pub trait Notifier: Send {
     fn notify_begin(&self);
+    fn notify_begin_diagnostics(&self);
+    fn notify_progress(&self, update: ProgressUpdate);
+    fn notify_publish_diagnostics(&self, PublishDiagnosticsParams);
     fn notify_end(&self);
-    fn notify_publish(&self, PublishDiagnosticsParams);
```

Comment:
I've reused this trait to now have notifications for both diagnostics and progress. I've renamed/reorded to make it clear what is what. However it got a bit messier than I wanted. For instance, I left the notifier implementation reference in the `PostBuildHandler`, which means to notify progress we need a reference to `pbh` in places like `cargo`, which isn't great.

Maybe separate into two traits?

The order of the current ones is:

* `notify_begin` starts progress
* `notify_progress` ... repeat for each update
* `notify_begin_diagnostics`
* `notify_publish_diagnostics`
* `notify_end` (ends both diagnostics and progress)

Reply:
It is unfortunate how much the `pbh` has to be propagated and would be nice to avoid that, but I don't see how we can do that.